pkg/elle/list_append/list_append.go

```diff
@@ -60,49 +208,359 @@ type rwExplainer struct {
+						break
+					}
+				}
+			}
+			return rwExplainResult{
+				Key:       k,
+				PreValue:  prev,
+				Value:     v,
+				AMopIndex: ai,
+				BMopIndex: i,
+			}
+		}
+	}
+	return nil
 }
 
-func (r *rwExplainer) RenderExplanation() string {
-	panic("implement me")
+func (r *rwExplainer) RenderExplanation(result core.ExplainResult, a, b string) string {
+	if result.Type() != core.RWDepend {
+		log.Fatalf("result type is not %s, type error", core.RWDepend)
+	}
+	er := result.(rwExplainResult)
+	key, prev, value := er.Key, er.PreValue, er.Value
+	if prev == nil {
+		return fmt.Sprintf("%s observed the initial (nil) state of %s, which %s created by appending %v",
+			a, key, b, value,
+		)
+	}
+	return fmt.Sprintf("%s did not observe %s's append of %v to %s",
+		a, b, value, key,
+	)
 }
 
 // rwGraph analyzes read-write anti-dependencies
-func rwGraph(history core.History) (core.Anomalies, core.DirectedGraph, core.DataExplainer) {
-	panic("implement me")
+func rwGraph(history core.History) (core.Anomalies, *core.DirectedGraph, core.DataExplainer) {
+	history, appendIdx, writeIdx, readIdx := preprocess(history)
+	g := core.NewDirectedGraph()
+
+	for _, op := range history {
+		for _, mop := range *op.Value {
+			if mop.IsAppend() {
+				deps := rwMopDeps(appendIdx, writeIdx, readIdx, op, mop)
+				var vdeps []core.Vertex
+				for dep := range deps {
+					vdeps = append(vdeps, core.Vertex{Value: dep})
+				}
+				g.LinkAllTo(vdeps, core.Vertex{Value: op}, core.RW)
+			}
+		}
+	}
+
+	return nil, g, &rwExplainer{
+		appendIdx: appendIdx,
+		writeIdx:  writeIdx,
+		readIdx:   readIdx,
+	}
 }
 
 // graph combines wwGraph, wrGraph and rwGraph
-func graph(history core.History) (core.Anomalies, core.DirectedGraph, core.DataExplainer) {
-	panic("implement me")
+func graph(history core.History) (core.Anomalies, *core.DirectedGraph, core.DataExplainer) {
+	return core.Combine(wwGraph, wrGraph, rwGraph)(history)
 }
 
-func g1aCases(history core.History) interface{} {
-	panic("implement me")
+type GCaseTp []core.Anomaly
+
+type G1Conflict struct {
+	Op      core.Op
+	Mop     core.Mop
+	Writer  core.Op
+	Element core.MopValueType
 }
 
-func g1bCases(history core.History) interface{} {
-	panic("implement me")
+func (g G1Conflict) IAnomaly() string {
+	return g.String()
 }
 
-func internalCases(history core.History) interface{} {
-	panic("implement me")
+func (g G1Conflict) String() string {
+	return fmt.Sprintf("(G1Conflict) Op: %s, mop: %s, writer: %s, element: %v", g.Op, g.Mop.String(), g.Writer.String(), g.Element)
 }
 
-func dirtyUpdateCases(history core.History) interface{} {
-	panic("implement me")
+func g1aCases(history core.History) GCaseTp {
+	failed := txn.FailedWrites(history)
+	okHistory := filterOkHistory(history)
+	iter := txn.OpMops(okHistory)
+	var excepted []core.Anomaly
+	for iter.HasNext() {
+		op, mop := iter.Next()
+		if mop.IsRead() {
+			versionMap := failed[mop.GetKey()]
+			v := mop.GetValue().(int)
+			writer := versionMap[v]
```

Comment:
if writer == nil, we should jump it, otherwise you will get a panic when deref writer on L327